### PR TITLE
nsfw image check

### PR DESCRIPTION
**Pull request by shivaylamba**
This pull request is related to issue #2 - Add support for disallowing people from uploading NSFW pictures with the help of [NSFWJS](https://www.npmjs.com/package/nsfwjs) library. 

Implementation: The idea is to use this client-side library to check the file being uploaded and verify whether the image has any NSFW content in it. The NSFW model predicts the class and if the predicted class has NSFW content, the file should be stopped from getting uploaded. 

The Upload.io [React Uploader](https://www.npmjs.com/package/react-uploader) didn't have a `onValidate` hook before and I got in touch with the founder @ljwagerfield who was super quick to add the `onValidate` hook and release the feature to production in a matter of a couple of hours (now that is quick product delivery!) 

Using the `onValidate` hook and the `NSFWPredictor.isSafeImg()` in case there is a file that is not safe, it will not get uploaded to UploadIo and the user will receive a message stating "Detected an NSFW image"

image

Previously before the `onValidate` hook was added to the Upload-io uploader, I was using a Tailwind Class to showcase an error for the detection of NSFW images. I can still add that if you think it is needed @Nutlope 

image

About the model: The ML model files (the model weights + model.json ) are currently hosted on an AWS S3 bucket, if needed I can share those with you and you could host them on Upload.io. Let me know if that works. The model URL is used in the `nsfwCheck.ts` file. 
```
'https://nsfw-model-1.s3.us-west-2.amazonaws.com/nsfw-predict-model/',
```

### CODE REVIEW

**Comment by ljwagerfield** [comment]
This is great work @shivaylamba 🎉 

We're looking into adding NSFW filtering server-side to Upload.io later this year, and this is an excellent stopgap solution.

Thank you again for the feature suggestion re. `onValidate: file => Promise<string | undefined>` for async custom file validation pre-upload 💪 

All the best!

File: pages/restore.tsx [hunk 1]
```diff
@@ -26,6 +27,15 @@ const options = {
   mimeTypes: ["image/jpeg", "image/png", "image/jpg"],
   editor: { images: { crop: false } },
   styles: { colors: { primary: "#000" } },
+  onValidate: async (file: any): Promise<undefined | string> => {
+    let isSafe = false;
+    try {
+      isSafe = await NSFWPredictor.isSafeImg(file);
+    } catch (error) {
+      console.log("NSFW predictor threw an error");
```
Comment (ljwagerfield) [inline pages/restore.tsx:35]:
```suggestion
      console.error("NSFW predictor threw an error", error);
```

File: pages/restore.tsx [hunk 2]
```diff
@@ -26,6 +27,15 @@ const options = {
   mimeTypes: ["image/jpeg", "image/png", "image/jpg"],
   editor: { images: { crop: false } },
   styles: { colors: { primary: "#000" } },
+  onValidate: async (file: any): Promise<undefined | string> => {
```
Comment (ljwagerfield) [inline pages/restore.tsx:30]:
```suggestion
  onValidate: async (file: File): Promise<undefined | string> => {
```

File: utils/nsfwCheck.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      throw error;
+    }
+  }
+
+  async classPredictor(file: File) {
+    const predictions = await this.predictImg(file, 3);
+    const neutralPrediction = predictions.find(
+      ({ className }) => className === 'Neutral'
+    );
+    console.log(predictions);
```
Comment (360macky) [inline utils/nsfwCheck.ts:74]: NIT: Hi @shivaylamba! There is a `console.log` in the `classPredictor` function

**Comment by llegomark** [comment]
I have tested the NSFW image check on my website, https://restorephotos.app/, and it appears to be functioning properly.
Screenshot (115)

**Comment by llegomark** [comment]
Web capture_21-1-2023_25156_github com
Web capture_21-1-2023_25226_github com
Web capture_21-1-2023_25210_github com

**Review by llegomark — CHANGES_REQUESTED** [review]
Great work on implementing the NSFW image check feature! However, I have noticed that there are some issues with false positives where pictures of white people or those that appear bright are being incorrectly tagged as NSFW. Can you please look into this and see if there is a way to improve the accuracy of the algorithm? Thank you!

**Comment by Nutlope** [comment]
Excellent work @shivaylamba, I'd like to merge this. Mind addressing the conflicts and we can get it merged?

**Comment by Nutlope** [comment]
This is now merged as a part of #24. Thanks all!

**Pull request closed without merging**